### DISCUSSION TITLE
Create individuals (and generate ontogeny factors) with custom ontogeny

### DISCUSSION THREAD

**Discussion by moPellowe** [Open-Systems-Pharmacology/Forum#1081] [Questions & Problems] [answered]
Is there a way to create individuals with a custom ontogeny as in PK-Sim, i.e. something along the lines of...
```
myProteinOntogeny <- MoleculeOntogeny$new(molecule = "MyProtein", ontogeny = "custom_ontogeny.csv")
individualCharacterstics <- createIndividualCharacteristics(
  species            = Species$Human,
  population         = HumanPopulation$Japanese_Population,
  gender             = Gender$Female,
  weight             = 75,
  height             = 1.75,
  heightUnit         = "m",
  age                = 43,
  moleculeOntogenies = myProteinOntogeny
)
```
?
I want to generate ontogeny factors based on my custom ontogeny for a population that already exists. So I was thinking that I would just create new individuals corresponding to the same demographics and copy over the ontogeny factors to my existing population.

**Comment by msevestre** [comment]
@moPellowe There is nothing like this implemented in the R toolbox. However, the only think that you need to do is add new "Variability Parameters" to your population

Say your protein is called "myProtein", then you will need to add those parameters

```
"myProtein|Ontogeny factor"                                               
"myProtein|Ontogeny factor GI"                                            
```

using population$setParameterValues (see doc here https://www.open-systems-pharmacology.org/OSPSuite-R/reference/Population.html#method-setParameterValues)

If you do not have separate ontogeny for GI vs rest of body, simply set the same values for both parameters

**Reply by moPellowe** [reply]
Oh thanks, @msevestre . Where can I find documentation on how PK-Sim generates ontogeny factors based on custom ontogeny tables?

**Reply by msevestre** [reply]
![image](https://user-images.githubusercontent.com/1041237/149319336-2ecad681-94b2-43ea-b2df-ef7ba34e7bab.png)

**Reply by moPellowe** [reply]
And how does PK-Sim interpolate for ages not explicitly provided in the table?

**Comment by msevestre** [comment, marked as answer]
Linear interpolation.
We should add a part in the doc describing this. ..feel free to submit a PR;-)

**Reply by StephanSchaller** [reply]
Enzyme/Transporter maturation ontogeny profiles are fitted using hill-equations, or are we talking about sth. else here?:
![image](https://user-images.githubusercontent.com/25477643/149361427-d55c92d1-52b7-4c71-a9ec-22bdc37d1b1f.png)

**Reply by StephanSchaller** [reply]
okay, now I get it. Well you could create a very dense ontogeny table based on the hill function fit and the linear interpolation will be fine.

**Reply by moPellowe** [reply]
My understanding is that Hill-equations were used for the built-in ontogenies and linear interpolation is used for custom ontogenies. Am I mistaken?

I'm trying to recreate how PK-Sim would handle custom ontogeny, so I think I'll leave the table as-is and use linear interpolation on that.

**Reply by prvmalik** [reply]
The hill equations are calculated/listed out in an excel table and it is those tables that are uploaded as the default ontogenies, not the equations. Still a table. You can download them from PKSim or look at the tables in the ontogeny documentation for these.

**Reply by msevestre** [reply]
Yes there is no hill equations involved in user defined ontogeny. Simple linear interpolation. Trust me on this

**Reply by StephanSchaller** [reply]
Yes, in the end it is always a table which gets interpolated, built-in or custom, but, if you want to create ontogeny profiles similar as the ones built-in, you first fit a hill equation and from it calculate a number of x/y combinations for your excel table. You then upload the table, only. Nevertheless, you could make this table very dense (i.e. write a lot of entries in your excel) so that the linear interpolation later will be more or less following the hill curve.

**Reply by moPellowe** [reply]
One follow-up question about the plotted ontogeny as in the figure above: is the grey area bounded by the 5th and 95th percentiles? It isn't explicitly said on the graph.

**Reply by Incei** [reply]
Dear @moPellowe,  grey area represents indeed the 5th and 95th percentiles, and are documented in the ontogeny documentation as prvmalik referred too. You can find the documentation here: https://github.com/Open-Systems-Pharmacology/OSPSuite.Documentation/blob/bc1df231fb1b8203e13e43c923b94d37b241cf04/PK-Sim%20Ontogeny%20Database%20Version%207.3.pdf

**Reply by Incei** [reply]
> Dear @moPellowe, grey area represents indeed the 5th and 95th percentiles, and are documented in the ontogeny documentation as prvmalik referred too. You can find the documentation here: https://github.com/Open-Systems-Pharmacology/OSPSuite.Documentation/blob/bc1df231fb1b8203e13e43c923b94d37b241cf04/PK-Sim%20Ontogeny%20Database%20Version%207.3.pdf

**Correction**: Dear @moPellowe,  the ontogeny table contains the Geo.Mean and Geo-SD! (The figure you referred to I cannot see, but the grey area of the fit-results using the Emax model in the ontogeny documentation (which I though you mentioned initially represents the 5th and 95th percentiles)

**Reply by moPellowe** [reply]
One other follow-up question. When reading in a table with a custom ontogeny into PK-Sim, how does PK-Sim calculate the 5th and 95th percentiles that are then plotted if it doesn't have the number of observations?

**Reply by msevestre** [reply]
It is not using percentile. As @Incei mentioned in his corrected answer, PKSim takes the values as is and assumes these values to be geo SD.
So the shades area are represented by min=mean/SD and max=mean*SD

**Reply by moPellowe** [reply]
So then a random value is then generated uniformly(?) between min = gmean/GSD and max = gmean*GSD?

**Reply by msevestre** [reply]
The random value is generated using a lognormal distribution with MEAN= gmean, and DEVIATION = gsd

**Reply by moPellowe** [reply]
So then for the interpolated ages, you would use a linearly interpolated gmean and a linearly interpolated gsd?

**Reply by msevestre** [reply]
yes.
Example of how the table look like for CYP3A4 in the PKSim DB if this helps

| molecule | species | group\_name | postmenstrual\_age | ontogeny\_factor | deviation |
| :--- | :--- | :--- | :--- | :--- | :--- |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 0.18 | 0.001 | 4.63 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 0.37 | 0.011 | 3.22 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 0.5 | 0.03 | 2.73 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 0.63 | 0.064 | 2.46 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 0.77 | 0.12 | 2.22 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 0.92 | 0.2 | 1.92 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 1.39 | 0.49 | 1.45 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 1.55 | 0.58 | 1.37 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 1.73 | 0.67 | 1.31 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 1.91 | 0.74 | 1.26 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 2.08 | 0.79 | 1.23 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 2.26 | 0.83 | 1.22 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 2.5 | 0.87 | 1.19 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 2.84 | 0.91 | 1.18 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 3.19 | 0.94 | 1.18 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 3.81 | 0.97 | 1.18 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 4.69 | 0.98 | 1.19 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 6.48 | 0.99 | 1.18 |
| CYP3A4 | Human | ONTOGENY\_LIVER\_NO\_GI | 11.17 | 1 | 1.18 |

**Reply by moPellowe** [reply]
Okay, yes, that makes sense. So if I was going to generate an ontogeny factor for an individual with age 2, I would then interpolate the gmean and GSD so that I would generate an ontogeny_factor from a log normal distribution with geomean 0.7664706 and GSD 1.244118 (which are the interpolated values for gmean and GSD from the entries for PMA=1.91 and 2.08) ?

**Reply by msevestre** [reply]
That's right except that  you would have to account for GA which is 40 weeks (here the values are PMA) so it would be age in years + 40/52 for a non preterm individual.
In your own script, since you will be adding variability to the population *by hand*, you do not have to do this.
